app/src/main/cpp/ExternalVR.cpp

```diff
@@ -420,6 +420,26 @@ ExternalVR::StopPresenting() {
   m.waitingForExit = true;
 }
 
+void
+ExternalVR::Shutdown() {
+  m.system.displayState.shutdown = true;
+  m.system.displayState.mSuppressFrames = true;
+  PushSystemState();
+
+  // Wait until Gecko has processed the shutdown message
+  // Otherwise the shmem data may be deallocated before Gecko processes it, leading to undefined behaviour
+  Wait wait(m.data.browserMutex, m.data.browserCond);
+  wait.Lock();
+  m.PullBrowserStateWhileLocked();
```

Comment:
This got me thinking about how to deal with the PullBrowserStateWhileLocked() issue. I wonder if we could use a closure to fix this issue where the code would look something like:
```
void PullBrowserState(std::function<bool(const BrowserState&)>& aFunc, float aWait = 0.0f) {
  wait.lock();
  copy_data();
  while(!aFunc(m.browserState)) {
    if (!wait.DoWait(aWait)) {
     break;
    }
    copy_data()
  }
}